espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ComposeNodeFinder.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.helpers
+
+import androidx.compose.ui.test.*
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidSelectorException
+import io.appium.espressoserver.lib.model.Strategy
+import io.appium.espressoserver.EspressoServerRunnerTest
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException
+import io.appium.espressoserver.lib.handlers.exceptions.StaleElementException
+import io.appium.espressoserver.lib.model.Locator
+import java.util.NoSuchElementException
+
+/**
+ * Retrieve cached node and return the SemanticsNodeInteraction
+ */
+@Throws(AppiumException::class)
+fun getNodeInteractionById(elementId: String?): SemanticsNodeInteraction {
+    elementId ?: throw InvalidArgumentException("Cannot find 'null' element")
+    val nodeInteraction = ComposeViewCache.get(elementId)
+        ?: throw StaleElementException(elementId)
+
+    try {
+        nodeInteraction.assertExists()
+    } catch (e: AssertionError) {
+        throw StaleElementException(elementId)
+    }
+    return nodeInteraction
+}
+
+/**
+ * Helper methods to find elements based on locator strategies and selectors
+ */
+fun findAllComposeNodeInteractions(
+    parentNode: SemanticsNodeInteraction?,
+    locator: Locator
+): SemanticsNodeInteractionCollection {
+    val matcher =
+        when (locator.using) {
+            Strategy.VIEW_TAG -> hasTestTag(locator.value!!)
+            Strategy.TEXT -> hasText(locator.value!!)
+            Strategy.LINK_TEXT -> hasText(locator.value!!)
+            Strategy.ACCESSIBILITY_ID -> hasContentDescription(locator.value!!)
+            else -> throw InvalidSelectorException(
+                "Can't use non-Compose selectors. " +
+                        "only ${Strategy.VIEW_TAG}, ${Strategy.TEXT}, ${Strategy.LINK_TEXT} and " +
+                        "${Strategy.ACCESSIBILITY_ID} are supported"
+            )
+        }
+
+    return if (parentNode == null) {
+        EspressoServerRunnerTest.composeTestRule.onAllNodes(matcher)
+    } else {
+        parentNode.onChildren().filter(matcher)
```

Comment:
shouldn't the parent node be added as well if matched?

Reply:
I did not get this. Added where?

Reply:
Added to the result. For example we have a hierarchy:
```
node[id='1']
    \node[id='11']
    \node[id='12']
```
Let say parent is `node[id='1']`, so then looking for parent.findElement('id=1') should return the parent element itself, whenever the API above would return no elements, because it only does matching on descendant nodes (e.g. `node[id='11']` and `node[id='12']` in thus case)

Reply:
`parent.findElement('id=1') should return the parent element itself`

This sounds a bit confusing. Suppose parent node has id = 1 and text 'hello'.  
`driver.find_element(id: 1).text => "hello"`

If we return the parent node itself in the result whenever the APU would return no descendant  elements, then this may happen:
```
@driver.find_element(id: 1).find_element(id: 1).text => "hello"
@driver.find_element(id: 1).find_element(id: 1).find_element(id: 1).text => "hello"
```

Reply:
this is how at least it works in WDA. What do you think @KazuCocoa ?

Reply:
I just checked in the existing espresso implementation. It returns only the descendants https://github.com/appium/appium-espresso-driver/blob/21ae67c6ae67d78aa068cc2c45d75c06015bc7cd/espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ViewFinder.kt#L263

Reply:
I don't mind to keep it as is unless we get any complains

Reply:
if the behavior is the same as espresso (which is the same package), I think of the compose also should be the same for now. (driver-wide behavior should be the same for now)
(Ideally, the behavior is supposed to be the same as other drivers as possible)